rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,775 @@
+    done in the following steps:
+    1. Optional: If the last dimension size of sequence- and sentence-level feature
+        differs, it's unified by applying a dense layer to each feature types.
+    1. sentence-level features are masked out everywhere except at the first available
+        position (`sequence_length+1`, where the length varies across input examples)
+    2. the padding of sequence-level features is increased by 1 to prepare empty space
+        for sentence-level features
+    3. sentence-level features are added just after the sequence-level ones
+
+    Arguments:
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        sequence_units: Last dimension size of the expected sequence-level features.
+        sentence_units: Last dimension size of the expected sentence-level features.
+        config: A model config for correctly parametrising the layer's components.
+
+    Input shape:
+        Tuple of three 3-D dense tensors, with shapes:
+            sequence: `(batch_size, max_seq_length, input_dim_seq)`
+            sentence: `(batch_size, 1, input_dim_sent)`
+            mask: `(batch_size, max_seq_length+1, 1)`
+
+    Output shape:
+        3-D tensor with shape: `(batch_size, sequence_length, units)` where:
+        - `units` matches `input_dim_seq` if that is the same as `input_dim_sent`,
+            otherwise `units = config[CONCAT_DIMENSION][attribute]`.
+        - `sequence_length` is the sum of 2nd dimension sizes of arguments `sequence`
+            and`sentence`.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        sequence_units: int,
+        sentence_units: int,
+        config: Dict[Text, Any],
+    ) -> None:
+        super().__init__(name=f"concatenate_sequence_sentence_features_{attribute}")
+
+        self._tf_layers = {}
+
+        # main use case -- both sequence and sentence features are expected
+        if sequence_units and sentence_units:
+            self.return_feature_type = f"{SEQUENCE}_{SENTENCE}"
+
+            # prepare dimension unifying layers if needed
+            if sequence_units != sentence_units:
+                self.output_units = config[CONCAT_DIMENSION][attribute]
+
+                for feature_type in [SEQUENCE, SENTENCE]:
+                    self._tf_layers[f"unify_dims.{feature_type}"] = layers.Ffnn(
+                        layer_name_suffix=f"unify_dims.{attribute}_{feature_type}",
+                        layer_sizes=[config[CONCAT_DIMENSION][attribute]],
+                        dropout_rate=config[DROP_RATE],
+                        reg_lambda=config[REGULARIZATION_CONSTANT],
+                        sparsity=config[WEIGHT_SPARSITY],
+                    )
+            else:
+                self.output_units = sequence_units
+
+        # edge cases where only sequence or only sentence features are expected
+        else:
+            if sequence_units and not entence_units:
```

Comment:
```suggestion
            if sequence_units and not sentence_units:
```

Reply:
Oh, yeah, I've noticed it too now... Thanks!